src/connector/src/sink/snowflake.rs

```diff
@@ -236,32 +229,106 @@ impl SnowflakeSinkWriter {
                 TimestamptzHandlingMode::UtcString,
                 TimeHandlingMode::String,
             ),
-            row_counter: 0,
-            payload: String::new(),
-            max_batch_row_num,
-            // initial value of `epoch` will start from 0
+            // initial value of `epoch` will be set to 0
             epoch: 0,
-        }
+            // will be (lazily) initialized after the begin of each epoch
+            // when some data is ready to be upload
+            streaming_uploader: None,
+        })
+    }
+
+    /// update the streaming uploader as well as the file suffix.
+    /// note: should *only* be called iff after a new epoch begins,
+    /// and `streaming_upload` being called the first time.
+    /// i.e., lazily initialization of the internal `streaming_uploader`.
+    async fn new_streaming_uploader(&mut self) -> Result<()> {
+        let file_suffix = self.file_suffix();
+        let path = generate_s3_file_name(self.s3_client.s3_path(), &file_suffix);
+        let uploader = self
+            .s3_client
+            .opendal_s3_engine
+            .streaming_upload(&path)
+            .await
+            .map_err(|err| {
+                SinkError::Snowflake(format!(
+                    "failed to create the streaming uploader of opendal s3 engine for epoch {}, error: {}",
+                    self.epoch,
+                    err
+                ))
+            })?;
+        self.streaming_uploader = Some((uploader, file_suffix));
+        Ok(())
     }
 
-    /// reset the `payload` and `row_counter`.
-    /// shall *only* be called after a successful sink to s3.
-    fn reset(&mut self) {
-        self.payload.clear();
-        self.row_counter = 0;
+    /// the `Option` is the flag to determine if there is data to upload.
+    fn has_data(&self) -> bool {
+        self.streaming_uploader.is_some()
     }
 
-    fn at_sink_threshold(&self) -> bool {
-        self.row_counter >= self.max_batch_row_num
+    /// write data to the current streaming uploader for this epoch.
+    async fn streaming_upload(&mut self, data: Bytes) -> Result<()> {
+        if !self.has_data() {
+            // lazily initialization
+            self.new_streaming_uploader().await?;
+        }
+        let Some(uploader) = self.streaming_uploader.as_mut() else {
+            return Err(SinkError::Snowflake(format!(
```

Comment:
It's better to unwrap here instead of returning error, because it's expected that the streaming uploader has been properly initialized when we reach here.

If we want to avoid the unwrap, the code can be changed to the following
```
let (uploader, _) = match &mut self.streaming_uploader {
  Some(t) => t,
  None => {
      let uploader = self.new_streaming_uploader().await?;
      self.streaming_uploader.insert(uploader)
  }
};
uploader
  .write_bytes(data)
...
```

with method `new_streaming_uploader` changed to a method simply returning the uploader without any side effect (modify the state of self). The signature shall be `async fn new_streaming_uploader(&mut self) -> Result<(Box<dyn StreamingUploader>, String)>`. (I had a try locally, though the method has no side effect, the code can only compile with &mut self. Using &self can pass the borrow checker, but will make the generated future not `Sync` )

Reply:
> with method new_streaming_uploader changed to a method simply returning the uploader without any side effect (modify the state of self).

I like the pure way to generate the new streaming uploader for each epoch - and we could then avoid the use of `unwrap`.